lite: Follow up from #3989

melekes opened the pull request:
Addresses comments in #3989 

- [x] https://github.com/tendermint/tendermint/pull/3989#discussion_r332751051
- [x] https://github.com/tendermint/tendermint/pull/3989#discussion_r352125455
- [x] https://github.com/tendermint/tendermint/pull/3989#pullrequestreview-315454506
- [x] cross-checking primary using alternative providers
- [x] https://github.com/tendermint/tendermint/pull/3989#discussion_r353117610 store only headers within the trusting period
- [x] more docs & tests

---

<!--

Thanks for filing a PR! Before hitting the button, please check the following items.
Please note that every non-trivial PR must reference an issue that explains the
changes in the PR.

-->

* [x] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

melekes commented:
> https://github.com/tendermint/tendermint/pull/3989#discussion_r332751051

if options.Height:

1) ahead of trustedHeader.Height => perform Backward verification
2) equals trustedHeader.Height => check Hash is the same (if not, remove trustedHeader and all the headers before)
3) behind trustedHeader.Height => remove all the headers between options.Height and trustedHeader.Height, then check the Hash is the same (if not, remove trustedHeader and all the headers before)

The intuition here is the user always right. I.e. if she decides to reset the light client with an older header, there must be a reason for it.

ancazamfir commented on lite2/verifier.go:
```diff
@@ -16,6 +16,15 @@ var (
+//	b) h2 is valid;
+//	c) either h2.ValidatorsHash equals h1NextVals.Hash()
+//		 OR trustLevel ([1/3, 1]) of last trusted validators (h1NextVals) signed
+//		 correctly  (if not, ErrNotEnoughVotingPowerSigned is returned);
```
IMO there should be two different errors for the two cases:
- the validator set of a skipped header cannot be trusted, i.e. <1/3rd of h1 validator set has signed (new error, something like `ErrValidatorSetCannotBeTrusted`)
- the validator set is trusted but < 2/3rds has signed (`ErrNotEnoughVotingPowerSigned`)

cwgoes replied:
I agree; I think errors should be as specific as possible, if we are checking a different conditional it should be a different error.

ancazamfir commented on lite2/client.go:
```diff
@@ -442,3 +613,62 @@ func (c *Client) compareNewHeaderWithRandomAlternative(h *types.SignedHeader) er
+		return
+	}
+
+	// 3) Remove all headers (except the very last one) that are outside of
```
So here we remove all expired headers in the store except the last one. This is also the one referred to by the client's `trustedHeader *types.SignedHeader`.
It is possible that there is a single expired header in the store, same as the client's `trustedHeader`. I believe any header verification from  this point on will fail as we do check for this in `Verify()`. 
But why do we need to keep it and not clean this up as well since we know it is not trusted anymore? If the client restarts with this store, it will restore this header only to be overwritten if `c.trustedHeader.Height != trustOptions.Height`. Also, shouldn't a client fail right away if initialized with an already expired header?

melekes replied:
> Also, shouldn't a client fail right away if initialized with an already expired header?

Yes.

> But why do we need to keep it and not clean this up as well since we know it is not trusted anymore?

because c.trustedHeader is used in VerifyHeader (both in bisection and linear verification). Removing it (setting c.trustedHeader to `nil`) can crash the light client process. We can remove the header from the store however. It should not affect the light client work.

cwgoes commented on lite2/client.go:
```diff
@@ -129,26 +144,111 @@ func NewClient(
+	var primaryHash []byte
+	switch {
+	case options.Height > c.trustedHeader.Height:
+		h, err := c.primary.SignedHeader(c.trustedHeader.Height)
```
This seems to mismatch the comment - shouldn't we fetch for `options.Height` ?

Also, why are we re-fetching a header we've already decided to trust - what's the point?

melekes replied:
> This seems to mismatch the comment - shouldn't we fetch for options.Height ?

We're fetching the header, which is trusted by the light client to ensure it's hash matches one from full node. If not, it (along with previous headers) can't be trusted, hence removed. 

That said, maybe we don't need them when the light client is reset with a newer header. Note I expose `Cleanup()` function to remove all the existing headers.

cwgoes replied:
Is there some way for the user to provide options such that the client will take the provided height and provided hash as *the* trusted header, regardless of whether it's in the future or not?

melekes replied:
> Is there some way for the user to provide options such that the client will take the provided height and provided hash as the trusted header, regardless of whether it's in the future or not?

this is the default behavior.

cwgoes commented on lite2/client.go:
```diff
@@ -129,26 +144,111 @@ func NewClient(
+	}
+
+	if !bytes.Equal(primaryHash, c.trustedHeader.Hash()) {
+		c.logger.Info("Prev. trusted header's hash %X doesn't match hash %X from primary provider",
```
Here is one place I think we might want the user to explicitly confirm that they wanted to do this.

cwgoes commented on lite2/client.go:
```diff
@@ -129,26 +144,111 @@ func NewClient(
+	case options.Height == c.trustedHeader.Height:
+		primaryHash = options.Hash
+	case options.Height < c.trustedHeader.Height:
+		// remove all the headers ( options.Height, trustedHeader.Height ]
```
Here is one place I think we might want the user to explicitly confirm that they wanted to do this.

cwgoes commented on lite2/client.go:
```diff
@@ -208,22 +333,25 @@ func (c *Client) TrustedHeader(height int64, now time.Time) (*types.SignedHeader
 	if err != nil {
 		return nil, err
 	}
+	if h == nil {
```
When would this happen - should we return an error?

melekes replied:
Someone requests outdated header (outside of trusting period - removed) OR header which is not yet verified (saved to `trustedStore`)? There's no error per se - light client just does not have the header.

cwgoes replied:
Hmm OK, as long as the caller handles `nil` correctly in all cases.

cwgoes commented on lite2/client.go:
```diff
@@ -442,3 +613,62 @@ func (c *Client) compareNewHeaderWithRandomAlternative(h *types.SignedHeader) er
+		h, err := c.trustedStore.SignedHeader(height)
+		if err != nil {
+			c.logger.Error("can't get a trusted header", "err", err, "height", height)
+			continue
```
Should we panic? When would this happen?

melekes replied:
> When would this happen?

problem with underlying database. no panics please. header will be removed during next iteration.

cwgoes commented on lite2/client.go:
```diff
@@ -442,3 +613,62 @@ func (c *Client) compareNewHeaderWithRandomAlternative(h *types.SignedHeader) er
+			c.logger.Error("can't get a trusted header", "err", err, "height", height)
+			continue
+		}
+		if h == nil {
```
Presumably, this will happen if we skipped headers?

Can we iterate through the actual keyspace instead perhaps?

melekes replied:
> Can we iterate through the actual keyspace instead perhaps?

Sure. But we'd have to add `Iterator` like function to `Store` interface.

cwgoes replied:
Hmm, is that not available? It's not a big deal I suppose.

cwgoes commented on lite2/client.go:
```diff
@@ -442,3 +613,62 @@ func (c *Client) compareNewHeaderWithRandomAlternative(h *types.SignedHeader) er
+		err = c.trustedStore.DeleteSignedHeaderAndNextValidatorSet(height)
+		if err != nil {
+			c.logger.Error("can't remove a trusted header & validator set", "err", err, "height", height)
+			continue
```
Seems like we should panic here?

melekes replied:
Panics are reserved for exceptional cases. This is not one of them. We'll retry delete once again after `c.removeNoLongerTrustedHeadersPeriod`.

ancazamfir commented:
I believe we should change a bit the code such that:
- bisection is called with a `newHeader` that has been verified with `verifyNewHeaderAndVals()` and  `newVals.VerifyCommit()`. These two can be combined in some new function `verifyUntrustedSignedHeader()`
- bisection calls `lastVals.VerifyCommitTrusting()` to check that `newVals` can be trusted
- If not, it retrieves the pivot and verifies it with `verifyUntrustedSignedHeader()`. If all good it goes into recursion.

The reasons are:
- `Verify()` can return early with `ErrTooMuchChange` (before `VerifyCommit()` is called), will trigger bisection only to discover when returning from the (potentially deep) recursive call that `Verify()` fails.
- we call `Verify()` multiple times with same `newHeader` and `newVals`. Even though this happens with different `lastHeader` and `lastVals`, there are checks in `Verify()` that are independent of the `last..` values.

Note that the above proposed changes are as described in the english spec.

ancazamfir commented:
We just had a discussion along the same lines for the rust version of lite client and @ebuchman  explained that it is in fact better to do the `newVals.VerifyCommit()` after bisection (the way it is done now in Go code) in order to prevent a lite client to perform a CPU-intensive full commit verification for a commit with thousands signatures it got from a bad node only to fail on `VerifyCommitTrusting()` later. 

As a follow up to this, we will update the spec.

melekes commented on lite2/client.go:
```diff
@@ -88,6 +90,15 @@ func AlternativeSources(providers []provider.Provider) Option {
 	}
 }
 
+// RemoveNoLongerTrustedHeadersPeriod option can be used to define how often
```
> I'm a little confused - do we expect the light client to be running as a daemon?

Yeah. Constantly running on people's servers or mobile phones or smth else

> Can this just be a flag or command to gaiacli [xyz]?

You mean making RemoveNoLongerTrustedHeadersPeriod optional. Totally possible, but most would probably enroll since it will be maintaining constant disk usage (opposite to storing all headers).

cwgoes replied:
Ah, more I mean "running every time gaiacli [xyz] is called" (if there are headers which are old).

melekes replied:
Got it. Will make this optional

melekes commented on lite2/client.go:
```diff
@@ -129,26 +144,111 @@ func NewClient(
+//     options.Height and trustedHeader.Height, then check the Hash is the same
+//     (if not, remove trustedHeader and all the headers before)
+//
+// The intuition here is the user always right. I.e. if she decides to reset
```
I'd have to think about the right amount of callbacks. Don't want to add too many to avoid callback hell.

cwgoes replied:
Yeah, that's fair. This could be split up into multiple functions perhaps.

melekes replied:
Added `ConfirmationFunction` option.

melekes commented:
@ancazamfir thanks for the update

melekes commented:
@cwgoes @ancazamfir I believe I've addressed most of your comments. Please review again. Note I'm planning to do https://github.com/tendermint/tendermint/pull/4209#issuecomment-570583895 in a follow up PR.

ancazamfir commented on lite2/verifier.go:
```diff
@@ -31,34 +40,34 @@ func Verify(
 				h1NextVals.Hash(),
 				h2.ValidatorsHash,
 			)
+			return ErrNewValSetCantBeTrusted{err}
```
Shouldn't we return a different error here?
```suggestion
			return ErrInvalidNewHeader{err}
```
I think the caller deals with this by checking if these are sequential headers and returning an error (vs for example continuing bisection in the skipping header case). But I don't see an advantage of returning the `ErrNewValSetCantBeTrusted` error type. Or maybe just add a comment.

melekes replied:
Removed all the new errors except ErrNewValSetCantBeTrusted

ancazamfir commented on lite2/verifier.go:
```diff
@@ -31,34 +40,34 @@ func Verify(
 		err := h1NextVals.VerifyCommitTrusting(chainID, h2.Commit.BlockID, h2.Height, h2.Commit, trustLevel)
 		if err != nil {
-			return err
+			return ErrNewValSetCantBeTrusted{err}
```
Looks like `VerifyCommitTrusting()` can return some errors that are fatal (the ones different than `ErrNotEnoughVotingPowerSigned`) and should not result in returning `ErrNewValSetCantBeTrusted`.

melekes replied:
Good catch! Thank you

ancazamfir commented on lite2/verifier.go:
```diff
@@ -31,34 +40,34 @@ func Verify(
 	err := h2Vals.VerifyCommit(chainID, h2.Commit.BlockID, h2.Height, h2.Commit)
 	if err != nil {
-		return err
+		return ErrNewHeaderCantBeTrusted{err}
```
Should we just return err here? Again assimilating with `ErrNewValSetCantBeTrusted` doesn't seem to be the right thing to do. IMO `ErrNewValSetCantBeTrusted` should only show on the skipping header case where we verify that > 1/3rd power of old validators have signed the new header, and should be the only case where bisection continues.

melekes replied:
> Again assimilating with ErrNewValSetCantBeTrusted doesn't seem to be the right thing to do.

NOTE these are 2 different errors - `ErrNewValSetCantBeTrusted` VS `ErrNewHeaderCantBeTrusted`

melekes replied:
Removed all new errors except ErrNewValSetCantBeTrusted

ancazamfir replied:
> NOTE these are 2 different errors - ErrNewValSetCantBeTrusted VS ErrNewHeaderCantBeTrusted

Ah! Sorry, my bad.

ancazamfir commented on lite2/client.go:
```diff
@@ -547,7 +547,7 @@ func (c *Client) bisection(
+		return errors.Wrapf(err, "failed to verify the header #%d", newHeader.Height)
 	}
 
 	if newHeader.Height == lastHeader.Height+1 {
```
Is this check still needed? Since it cannot happen that we get `ErrNewValSetCantBeTrusted` with adjacent headers this could be handled by the `default:` above.

melekes replied:
Nope. Removed.

ancazamfir commented on lite2/verifier.go:
```diff
@@ -54,20 +54,25 @@ func Verify(
-			return ErrNewValSetCantBeTrusted{err}
+			switch e := err.(type) {
+			case types.ErrNotEnoughVotingPowerSigned:
+				return ErrNewValSetCantBeTrusted{e}
```
I might be missing something but I don't understand why not just return err, why do we need translation for this case? Couldn't `VerifyCommitTrusting()` return `ErrNewValSetCantBeTrusted` if `trustLevel` was not achieved?

melekes replied:
> Couldn't VerifyCommitTrusting() return ErrNewValSetCantBeTrusted if trustLevel was not achieved?

Nope because it does not know that we're checking new validator set. It only knows that vals presence in commit is some number.

ancazamfir commented:
Unrelated to your changes, why does `validator_set.go:verifyCommitBasic()` have `ValidatorSet` as target? It never uses 'vals'.

melekes commented:
> Unrelated to your changes, why does validator_set.go:verifyCommitBasic() have ValidatorSet as target? It never uses 'vals'.

another good catch. will make it pure function
